sympy/polys/matrices/sdm.py

```diff
@@ -1413,6 +1526,261 @@ def sdm_irref(A):
+    :ref:`ZZ`, :ref:`QQ` and :ref:`QQ(a)`. In the case of :ref:`QQ` or
+    :ref:`K(x)` it might be more efficient to clear denominators and use
+    :ref:`ZZ` or :ref:`K[x]` instead.
+
+    For inexact domains like :ref:`RR` and :ref:`CC` use ``ddm_irref`` instead.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.matrices.sdm import sdm_rref_den
+    >>> from sympy.polys.domains import ZZ
+    >>> A = {0: {0: ZZ(1), 1: ZZ(2)}, 1: {0: ZZ(3), 1: ZZ(4)}}
+    >>> A_rref, den, pivots = sdm_rref_den(A, ZZ)
+    >>> A_rref
+    {0: {0: -2}, 1: {1: -2}}
+    >>> den
+    -2
+    >>> pivots
+    [0, 1]
+
+    See Also
+    ========
+
+    sympy.polys.matrices.domainmatrix.DomainMatrix.rref_den
+        Higher-level interface to ``sdm_rref_den`` that would usually be used
+        instead of calling this function directly.
+    sympy.polys.matrices.sdm.sdm_rref_den
+        The ``SDM`` method that uses this function.
+    sdm_irref
+        Computes RREF using field division.
+    ddm_irref_den
+        The dense version of this algorithm.
+
+    References
+    ==========
+
+    .. [1] Fraction-free algorithms for linear and polynomial equations.
+        George C. Nakos , Peter R. Turner , Robert M. Williams.
+        https://dl.acm.org/doi/10.1145/271130.271133
+    """
+    #
+    # We represent each row of the matrix as a dict mapping column indices to
+    # nonzero elements. We will build the RREF matrix starting from an empty
+    # matrix and appending one row at a time. At each step we will have the
+    # RREF of the rows we have processed so far.
+    #
+    # Our representation of the RREF divides it into three parts:
+    #
+    # 1. Fully reduced rows having only a single nonzero element (the pivot).
+    # 2. Partially reduced rows having nonzeros after the pivot.
+    # 3. The current denominator and divisor.
+    #
+    # When adding a new row we need to multiply it by the current denominator.
+    # Then we reduce the new row by cross cancellation with the previous rows.
+    # Then if it is not reduced to zero we take its leading entry as the new
+    # pivot, cross cancel the new row from the previous rows and update the
+    # denominator. In the fraction-free version this last step requires
+    # multiplying and dividing the whole matrix by the new pivot and the
+    # current divisor. The advantage of building the RREF one row at a time is
+    # that in the sparse case we only need to work with the relatively sparse
+    # upper rows of the matrix. The simple version of FFGJ in [1] would
+    # multiply and divide all the dense lower rows at each step.
+
+    # Handle the trivial cases.
+    if not A:
+        return ({}, K.one, [])
+    elif len(A) == 1:
+        Ai, = A.values()
+        j = min(Ai)
+        Aij = Ai[j]
+        return ({0: Ai.copy()}, Aij, [j])
+
+    # Make sure we have the rows in order to make this deterministic from the
```

Comment:
What you are doing here is what I was thinking would make the denominator canonical. I can't see the whole picture clearly, but is the non-canonical denominator still an issue (in which case the note in the tests can be deleted) or can this approach be used somewhere else to address the concern that you had raised?

Reply:
This line sorts the dictionary of rows so that we get a list of rows in top-to-bottom order for the matrix. That means that this function does not depend on the ordering of the internal dictionary keys and so is deterministic. Otherwise the result might be different for say `DomainMatrix({1:{0:1}, 0: {0:1}})` and `DomainMatrix({0:{0:1}, 1: {0:1}})` even though these both represent the exact same matrix.

The part about the denominator being canonical is different. The sparse and dense versions of the algorithm can return results with different denominators even for the exact same matrix. Also both implementations can return results with different denominators depending on the order of the rows of a matrix even though ordinary RREF is invariant under any permutation of the rows e.g.:
```python
In [27]: M = randMatrix(5, 10, percent=20)

In [28]: print(repr(M))
Matrix([
[0, 26,  0,  0,  0, 66, 17, 0,  0, 0],
[0,  0,  0, 94,  0,  0,  0, 0, 31, 0],
[0,  0, 29,  0,  0,  0,  0, 0,  0, 0],
[0,  0,  0,  0,  0,  0,  0, 0,  0, 0],
[0, 92,  0,  0, 73,  0,  0, 0, 49, 0]])

In [29]: print(repr(M.rref()[0]))
Matrix([
[0, 1, 0, 0, 0,     33/13,    17/26, 0,     0, 0],
[0, 0, 1, 0, 0,         0,        0, 0,     0, 0],
[0, 0, 0, 1, 0,         0,        0, 0, 31/94, 0],
[0, 0, 0, 0, 1, -3036/949, -782/949, 0, 49/73, 0],
[0, 0, 0, 0, 0,         0,        0, 0,     0, 0]])

In [30]: print(repr(M[::-1,:].rref()[0])) # reverse the rows
Matrix([
[0, 1, 0, 0, 0,     33/13,    17/26, 0,     0, 0],
[0, 0, 1, 0, 0,         0,        0, 0,     0, 0],
[0, 0, 0, 1, 0,         0,        0, 0, 31/94, 0],
[0, 0, 0, 0, 1, -3036/949, -782/949, 0, 49/73, 0],
[0, 0, 0, 0, 0,         0,        0, 0,     0, 0]])
```
With `rref_den` a different order might result in a different denominator e.g. here the sign is flipped:
```python
In [33]: print(repr(M.to_DM().rref_den()[0].to_Matrix()))
Matrix([
[0, 5173948,       0,       0,       0,  13133868,  3382966, 0,       0, 0],
[0,       0, 5173948,       0,       0,         0,        0, 0,       0, 0],
[0,       0,       0, 5173948,       0,         0,        0, 0, 1706302, 0],
[0,       0,       0,       0, 5173948, -16552272, -4263464, 0, 3472924, 0],
[0,       0,       0,       0,       0,         0,        0, 0,       0, 0]])

In [34]: print(repr(M[::-1,:].to_DM().rref_den()[0].to_Matrix()))
Matrix([
[0, -5173948,        0,        0,        0, -13133868, -3382966, 0,        0, 0],
[0,        0, -5173948,        0,        0,         0,        0, 0,        0, 0],
[0,        0,        0, -5173948,        0,         0,        0, 0, -1706302, 0],
[0,        0,        0,        0, -5173948,  16552272,  4263464, 0, -3472924, 0],
[0,        0,        0,        0,        0,         0,        0, 0,        0, 0]])
```